numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1553,13 +1558,25 @@ pack_inner(const char *inptr,
         if (index == n_out - 1) {
             build <<= 8 - remain;
         }
+        if (order == 'l') {
+            char tmp = 0;
+            tmp |= build &   1 ? 128 : 0;
+            tmp |= build &   2 ?  64 : 0;
+            tmp |= build &   4 ?  32 : 0;
+            tmp |= build &   8 ?  16 : 0;
+            tmp |= build &  16 ?   8 : 0;
+            tmp |= build &  32 ?   4 : 0;
+            tmp |= build &  64 ?   2 : 0;
+            tmp |= build & 128 ?   1 : 0;
+            build = tmp;
+        }
```

Comment:
There probably is a more elegant way to flip the bits

Reply:
I think you need to rewrite the loop about, and not do `build <<= 1`, but rather OR-in an already shifted element (i.e., something like `build |= (inputr[j] != 0) << shift` where shift depends on `order` and `j`).

Reply:
adopted

Reply:
You mention "adopted" but I think the code is still the same, no? I meant the loop *above* (not "about"), which now has the `build <<= 1`. I think that loop can be rewritten to just do the right thing for each order.

Reply:
I could write this as below, is that better? I don't understand why I needed to use `unsigned char` instead of `char` for the `order != 'b'` case, I am only ever shifting right. If `unsigned char` is required it should be for the left shifts, no?

<details>

```
    for (; index < n_out; index++) {
        unsigned char build = 0;
        int i, maxi;
        npy_intp j;

        maxi = (index == n_out - 1) ? remain : 8;
        if (order == 'b') {
            for (i = 0; i < maxi; i++) {
                build <<= 1;
                for (j = 0; j < element_size; j++) {
                    build |= (inptr[j] != 0);
                }
                inptr += in_stride;
            }
            if (index == n_out - 1) {
                build <<= 8 - remain;
            }
        }
        else
        {
            for (i = 0; i < maxi; i++) {
                build >>= 1;
                for (j = 0; j < element_size; j++) {
                    build |= (inptr[j] != 0) ? 128 : 0;
                }
                inptr += in_stride;
            }
            if (index == n_out - 1) {
                build >>= 8 - remain;
            }
        }
        *outptr = (char)build;
        outptr += out_stride;
    }

```